### ISSUE TITLE
Blazor WebSockets can't handle Chrome V8 debugger

### ISSUE THREAD

**Issue by kumpera** [aspnet/Home#2786]
I'm using dotnet core 2.1.4 from OSX High Sierra.

When trying to run the Blazor sample that features a V8 protocol proxy, it leads to frequent disconnects from the browser side. It doesn't happen if I use a simple sockets proxy or direct connection.

To reproduce:
Clone https://github.com/SteveSanderson/Blazor
Switch to the uses-dna branch
Build it with `dotnet build Blazor.Core.sln`
Run the server example with: `cd samples/ClientServerApp/ClientServerApp.Server; dotnet run``
Start chrome with remote debugging enabled: `/Applications/Google\ Chrome.app/Contents/MacOS/Google\ Chrome --remote-debugging-port=9222`
Open a tab and browser to `localhost:5000`
Open a tab and browser to `localhost:9223`
Select an available tab to debug.
Leave it idle for around 5-30 seconds and see it disconnect all the time.

This was an issue originally in corefx: https://github.com/dotnet/corefx/issues/26369

**Comment by natemcmaster** [comment]
Blazor isn't an official ASP.NET framework. At the moment, it's just an experiment with webassembly that @SteveSanderson has been investigating. WebSockets and many .NET APIs do not work in this experiment.

For context: 
http://blog.stevensanderson.com/2017/11/05/blazor-on-mono/
https://github.com/SteveSanderson/Blazor#why-isnt-there-an-issue-tracker-how-do-i-file-issues

**Comment by kumpera** [comment]
I @natemcmaster, it's fine, I'm not asking you to support Blazor, I'm providing you a test case so you can reproduce my issue.

**Comment by kumpera** [comment]
Blazor, like any other project using asp.net core, doesn't implement WebSockets itself.

**Comment by kumpera** [comment]
Hey @natemcmaster this is a blocking issue to your coworkers in the mono team. It's blocking us from using asp.net core.

**Comment by Eilon** [comment]
@kumpera - can you email me and Dan and Steve about this?

**Comment by Eilon** [comment]
This issue was moved to aspnet/Blazor#42

**Comment by Eilon** [comment]
@anurse - any idea on who the right contact is for V8 debugger + WebSockets?

**Comment by kumpera** [comment]
Hey guys,

Sorry for the lousy repro. Here's the simplest one I could come up with that shows the issue:
https://github.com/kumpera/websockets-issue

I have strong reasons to believe this is an issue with asp.net as I managed to get such a proxy working with multiple alternative websockets implementations.

**Comment by Eilon** [comment]
Thanks @kumpera .

**Comment by anurse** [comment]
> @anurse - any idea on who the right contact is for V8 debugger + WebSockets?

Sundar Pichai? But seriously no.

**Comment by Eilon** [comment]
Well, for the .NET Core Web Sockets implementation that would be used in this scenario. Is it ASP.NET Core's implementation, or a .NET Core one?

**Comment by danroth27** [comment]
Folks, the issue being reported here is with the websockets support in ASP.NET Core. It's being used to setup a development server process for Blazor, but the issue is actually server side.

**Comment by anurse** [comment]
Looking at your sample, the reason you aren't seeing any errors is because you never observe the exception in the two "ProxyMsg" tasks you start up. When I added code to observe the exceptions I saw this error:

```
fail: repro.Startup[0]
      Error processing browser socket
System.Net.WebSockets.WebSocketException (0x80004005): The remote party closed the WebSocket connection without completing the close handshake. ---> System.Net.WebSockets.WebSocketException (0x80004005): The remote party closed the WebSocket connection without completing the close handshake.
   at System.Net.WebSockets.ManagedWebSocket.<EnsureBufferContainsAsync>d__70.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Net.WebSockets.ManagedWebSocket.<ReceiveAsyncPrivate>d__61.MoveNext()
   at System.Net.WebSockets.ManagedWebSocket.<ReceiveAsyncPrivate>d__61.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at repro.Startup.<ProxyMsg>d__1.MoveNext() in /Users/anurse/Code/kumpera/websockets-issue/Startup.cs:line 35
got exception System.Net.WebSockets.WebSocketException (0x80004005): The remote party closed the WebSocket connection without completing the close handshake. ---> System.Net.WebSockets.WebSocketException (0x80004005): The remote party closed the WebSocket connection without completing the close handshake.
   at System.Net.WebSockets.ManagedWebSocket.<EnsureBufferContainsAsync>d__70.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Net.WebSockets.ManagedWebSocket.<ReceiveAsyncPrivate>d__61.MoveNext()
   at System.Net.WebSockets.ManagedWebSocket.<ReceiveAsyncPrivate>d__61.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at repro.Startup.<ProxyMsg>d__1.MoveNext() in /Users/anurse/Code/kumpera/websockets-issue/Startup.cs:line 60
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at repro.Startup.<>c__DisplayClass2_0.<<Configure>b__1>d.MoveNext() in /Users/anurse/Code/kumpera/websockets-issue/Startup.cs:line 95
```

So the request shuts down because the browser is terminating it's connection without performing a close handshake. So it appears the disconnect is originating from Chrome. That socket is a `ClientWebSocket` so it's not directly related to ASP.NET as far as I can tell. Happy to take another look if you can get more data as to why the browser is disconnecting abnormally.

**Comment by kumpera** [comment]
@anurse I understand that, what I don't get it is why a sockets proxy doesn't experience this.
Which is what makes me believe this to be an issue with ASP.NET WebSockets.

Does that make sense?

**Comment by danroth27** [comment]
@SteveSandersonMS

**Comment by anurse** [comment]
@kumpera I can't really answer that question, I'm not familiar with Chrome's remote debugging implementation. You'd need to look at network traces to get more information about what's going on.

**Comment by kumpera** [comment]
@anurse err, why should I be debugging WebSockets communication?
I'm not a network engineer and this is a bug in the asp.net stack.

Let me rephrase my issue to see if it makes sense to you.
Is ASP.NET WebSockets code supposed to be able to handle proxying WebSockets?
Should it be able to be behaviorally equivalent to what a sockets-based proxy would accomplish?

**Comment by kumpera** [comment]
@anurse btw, this issue was originally opened on CoreFx and they told me that it was an asp.net issue:
https://github.com/dotnet/corefx/issues/26369

**Comment by kumpera** [comment]
@anurse And it worked! Thanks a lot!

Btw, would it make sense to make those ping messages available in the API?

**Comment by kumpera** [comment]
Closing this issue as the proposed solution works great.

Thanks again for the help :)

**Comment by anurse** [comment]
> Btw, would it make sense to make those ping messages available in the API?

That's a question for dotnet/corefx. It might make some sense, but it's going to be difficult as it would be a pretty significant breaking change.
